reverse_proxy_handler.cc

```diff
@@ -0,0 +1,142 @@
+#include <string>
+#include <memory>
+
+#include "request_handler.h"
+#include "reverse_proxy_handler.h"
+#include "response.h"
+#include "config_parser.h"
+#include "http_client.h"
+
+#include <algorithm>
+
+#include <boost/log/trivial.hpp>
+
+RequestHandler::Status ReverseProxyHandler::Init(const std::string& uri_prefix, const NginxConfig& config)
+{
+	prefix_ = uri_prefix;
+
+	for (auto statement : config.statements_)
+	{
+		if(statement->tokens_.size() == 2 && statement->tokens_[0] == "proxy_pass")
+		{
+			url_ = statement->tokens_[1];
+
+			std::string::size_type protocol_pos = url_.find("//");
+			if(protocol_pos == std::string::npos) {
+				BOOST_LOG_TRIVIAL(error) << "proxy_pass didn't specify protocol.'";
+				return RequestHandler::ERROR;
+			}
+			protocol_ = url_.substr(0, protocol_pos - 1);
+
+			std::string::size_type host_pos = url_.find('/', protocol_pos + 2);
+			if(host_pos != std::string::npos) {
+				host_ = url_.substr(protocol_pos + 2, host_pos - protocol_pos - 2);
+				path_ = url_.substr(host_pos);
+			}
+			else {
+				host_ = url_.substr(protocol_pos + 2);
+				path_ = "/";
+			}
+			printf("path_: %s\n", path_.c_str());
+			return RequestHandler::OK;
+		}
+	}
+
+	BOOST_LOG_TRIVIAL(error) << "proxy_pass not specified in static handler for " << uri_prefix;
+	return RequestHandler::ERROR;
+}
+
+RequestHandler::Status ReverseProxyHandler::HandleRequest(const Request& request, Response* response) const
+{
+	BOOST_LOG_TRIVIAL(trace) << "Creating Reverse Proxy Response...";
+
+	Request OutgoingRequest = TransformIncomingRequest(request);
+	
+	std::unique_ptr<Response> resp;
+	
+	std::string new_host = host_;
+	std::string new_uri = path_;
+	for(int i = 0; i < MaxRedirectDepth; i++) {
+		BOOST_LOG_TRIVIAL(trace) << "Reaching out to " << new_host << " with URI: " << new_uri;
+		resp.reset(); // we don't need the old response
+		resp = VisitOutsideServer(OutgoingRequest, new_host, protocol_);
+		if(resp == nullptr) {
+			return RequestHandler::ERROR;
+		}
+		BOOST_LOG_TRIVIAL(trace) << "response received";
+		if(resp->GetStatusCode() == Response::ResponseCode::found) {
+			std::string url = resp->get_header("Location");
+			if(url == "") { //error, response syntax doesn't tell us where to go
+				BOOST_LOG_TRIVIAL(error) << "Error in redirect received, location to go to not specified";
+				return RequestHandler::ERROR;
+			}
+			
+
+
+			std::string::size_type protocol_pos = url.find("//");
+			if(protocol_pos == std::string::npos) {
+				BOOST_LOG_TRIVIAL(error) << "response didn't specify protocol.'";
+				return RequestHandler::ERROR;
+			}
+
+			std::string::size_type host_pos = url.find('/', protocol_pos + 2);
+			if(host_pos != std::string::npos) {
+				new_host = url.substr(protocol_pos + 2, host_pos - protocol_pos - 2);
+				new_uri = url.substr(host_pos);
+			}
+			else {
+				new_host = url.substr(protocol_pos + 2);
+				new_uri = "/";
+			}
```

Comment:
Yeah, this may be a good idea, but probably out of the scope of the config parser. We can implement this in the future.